### PR TITLE
refactor: vault should be slashed first on theft

### PR DESCRIPTION
**Pull request by gregdhill**
Signed-off-by: Gregory Hill <gregorydhill@outlook.com>

### CODE REVIEW

**Comment by gregdhill** [comment]
Let's look at an example,

```
collateral_tokens = issued_tokens + to_be_issued_tokens
liquidated_collateral = min(backing_collateral, (collateral_tokens * exchange_rate) * secure_collateral_threshold)
collateral_to_slash_now = (liquidated_collateral * (collateral_tokens - to_be_redeemed_tokens)) / collateral_tokens
collateral_for_to_be_redeemed = liquidated_collateral - collateral_to_slash_now
```

In the "vault first slashing" proposal, `collateral_to_slash_now` is subtracted from Vault's collateral first and the remaining amount (if any) is slashed from the staking pool (all nominators proportionally). The `collateral_for_to_be_redeemed` is then temporarily slashed from the remaining participants and re-distributed upon successful redeem.

```
actual_stake = stake - (stake * (slashed_amount / total_stake))
```

# Examples

```
exchange_rate = 1 (i.e. 1 BTC = 1 DOT)
secure_collateral_threshold = 1.5 (150%)
issued_tokens = 100
to_be_issued_tokens = 0
to_be_redeemed_tokens = 20
```

## Example 1 - Overcollateralized (e.g. Theft)

**Before Slashing**
``` 
vault_collateral = 200
nominator_collateral = 20
```

**During Slashing**
```
liquidated_collateral = min(220, ((100 + 0) * 1) * 1.5) = 150
collateral_to_slash_now = (150 * (100 - 20)) / 100 = 120
collateral_for_to_be_redeemed = 150 - 120 = 30
```

**After First Slashing**
```
vault_collateral = 80
nominator_collateral = 20
```

**After Second Slashing**
```
vault: 80 - (80 * (30 / 100)) = 56
nominator: 20 - (20 * (30 / 100)) = 14
```

Assuming all redeems are completed successfully, state is as in after first slashing.

## Example 2 - Undercollateralized

**Before Slashing**
``` 
vault_collateral = 90
nominator_collateral = 10
```

**During Slashing**
```
liquidated_collateral = min(100, ((100 + 0) * 1) * 1.5) = 100
collateral_to_slash_now = (100 * (100 - 20)) / 100 = 80
collateral_for_to_be_redeemed = 100 - 80 = 20
```

**After First Slashing**
```
vault_collateral = 10
nominator_collateral = 10
```

**After Second Slashing**
```
vault: 10 - (10 * (20 / 20)) = 0
nominator: 10 - (10 * (20 / 20)) = 0
```

Assuming all redeems are completed successfully, state is as in after first slashing.

**Comment by nud3l** [comment]
The vault first and proportional is almost the same. Just when the to_be_redeemed_tokens are different, you'll note a difference. Slightly adjusted your example above:

# Example 2 - Undercollateralized (modified)

```
exchange_rate = 1 (i.e. 1 BTC = 1 DOT)
secure_collateral_threshold = 1.5 (150%)
issued_tokens = 100
to_be_issued_tokens = 0
to_be_redeemed_tokens = 10
```

**Before Slashing**

```
vault_collateral = 90
nominator_collateral = 10
```

**During Slashing**

```
liquidated_collateral = min(100, ((100 + 0) * 1) * 1.5) = 100
collateral_to_slash_now = (100 * (100 - 10)) / 100 = 90
collateral_for_to_be_redeemed = 100 - 90 = 10
```

**After First Slashing**

```
vault_collateral = 0
nominator_collateral = 10
```

**After Second Slashing**

```
vault: 10 - (10 * (20 / 20)) = 0
nominator: 10 - (10 * (20 / 20)) = 0
```

Assuming all redeems are completed successfully, state is as in after first slashing. In this example, the Vault does not have any incentive to fulfill the redeem request any longer since all its collateral has been slashed already.

**Comment by gregdhill** [comment]
## Example 2 - Undercollateralized (Proportional Slashing)

**Before Slashing**
``` 
vault_collateral = 90
nominator_collateral = 10
```

**During Slashing**
```
liquidated_collateral = min(100, ((100 + 0) * 1) * 1.5) = 100
collateral_to_slash_now = (100 * (100 - 20)) / 100 = 80
collateral_for_to_be_redeemed = 100 - 80 = 20
```

**After First Slashing**
```
vault_collateral = 90 - (90 * (80 / 100)) = 18
nominator_collateral = 10 - (10 * (80 / 100)) = 2
```

**After Second Slashing**
```
vault: 18 - (18 * (20 / 20)) = 0
nominator: 2 - (2 * (20 / 20)) = 0
```

Assuming all redeems are completed successfully, state is as in after first slashing.

File: crates/nomination/src/lib.rs [hunk 1]
```diff
@@ -292,8 +292,8 @@ impl<T: Config> Pallet<T> {
     }
 
     pub fn get_total_nominated_collateral(vault_id: &T::AccountId) -> Result<Collateral<T>, DispatchError> {
-        let vault_backing_collateral = ext::vault_registry::get_backing_collateral::<T>(vault_id)?;
-        let vault_actual_collateral = ext::vault_registry::compute_collateral::<T>(vault_id)?;
+        let vault_backing_collateral = ext::vault_registry::get_vault_total_collateral::<T>(vault_id)?;
+        let vault_actual_collateral = ext::vault_registry::get_vault_collateral::<T>(vault_id)?;
```
Comment (nud3l) [inline crates/nomination/src/lib.rs:296]:
```suggestion
        let vault_only_collateral = ext::vault_registry::get_vault_collateral::<T>(vault_id)?;
```

Maybe that makes it a bit more explicit?

File: crates/nomination/src/ext.rs [hunk 2]
```diff
@@ -39,8 +39,16 @@ pub(crate) mod vault_registry {
     pub use frame_support::dispatch::{DispatchError, DispatchResult};
     pub use vault_registry::{DefaultVault, VaultStatus};
 
-    pub fn get_backing_collateral<T: crate::Config>(vault_id: &T::AccountId) -> Result<Collateral<T>, DispatchError> {
-        <vault_registry::Pallet<T>>::get_backing_collateral(vault_id)
+    // excluding nomination
+    pub fn get_vault_collateral<T: crate::Config>(vault_id: &T::AccountId) -> Result<Collateral<T>, DispatchError> {
```
Comment (nud3l) [inline crates/nomination/src/ext.rs:43]:
```suggestion
    pub fn get_vault_only_collateral<T: crate::Config>(vault_id: &T::AccountId) -> Result<Collateral<T>, DispatchError> {
```
Maybe more explicit that way?

**Comment by gregdhill** [comment]
Vault first slashing:

- Leaves Nominators untouched when Vault is sufficiently overcollateralized
- Reduced incentive to execute redeems if undercollateralized

Proportional slashing:

- Nominators are also slashed on overcollateralization
- More incentive to execute redeems if undercollateralized

**Comment by nud3l** [comment]
**Summary**

**Comment by gregdhill** [comment]
## Example 1 - Overcollateralized (Reverse)

**Before Slashing**
``` 
vault_collateral = 200
nominator_collateral = 20
```

**During Slashing**
```
liquidated_collateral = min(220, ((100 + 0) * 1) * 1.5) = 150
collateral_to_slash_now = (150 * (100 - 20)) / 100 = 120
collateral_for_to_be_redeemed = 150 - 120 = 30
```

**After First Slashing (prev. Second)**
```
vault_collateral = 200 - 30 = 170
nominator_collateral = 20
```

**After Second Slashing (prev. First) - Vault First**
```
vault: 170 - 120 = 50
nominator: 20
```

**After Second Slashing (prev. First) - Proportional**
```
vault: 170 - (170 * (120 / 190)) = ~62
nominator: 20 - (20 * (120 / 190)) = ~7
```

## Example 2 - Undercollateralized (Reverse)

**Before Slashing**
``` 
vault_collateral = 90
nominator_collateral = 10
```

**During Slashing**
```
liquidated_collateral = min(100, ((100 + 0) * 1) * 1.5) = 100
collateral_to_slash_now = (100 * (100 - 20)) / 100 = 80
collateral_for_to_be_redeemed = 100 - 80 = 20
```

**After First Slashing (prev. Second)**
```
vault_collateral = 90 - 20 = 70
nominator_collateral = 10
```

**After Second Slashing (prev. First)**
```
vault: 70 - (70 * (80 / 80)) = 0
nominator: 10 - (10 * (80 / 80)) = 0
```

**Comment by gregdhill** [comment]
This now implement vault first slashing with the adaption that the collateral for outstanding to-be-redeemed tokens is slashed first. This means we no longer need to "unslash" since the collateral is credited to the vault.

File: crates/staking/src/lib.rs [hunk 3]
```diff
@@ -317,7 +317,9 @@ impl<T: Config> Pallet<T> {
     ) -> Result<(), DispatchError> {
         let nonce = Self::nonce(currency_id, vault_id);
         let total_stake = Self::total_stake_at_index(currency_id, (nonce, vault_id));
-        if total_stake.is_zero() {
+        if amount.is_zero() {
```
Comment (sander2) [inline crates/staking/src/lib.rs:320]: Make sure to update spec

File: crates/vault-registry/src/types.rs [hunk 4]
```diff
@@ -444,7 +447,20 @@ impl<T: Config> RichVault<T> {
     }
 
     pub(crate) fn slash_to_liquidation_vault(&mut self, amount: Collateral<T>) -> DispatchResult {
-        ext::staking::slash_stake::<T>(&self.id(), amount)?;
+        let vault_id = self.id();
+
+        // get the collateral supplied by the vault (i.e. excluding nomination)
+        let collateral = Pallet::<T>::compute_collateral(&vault_id)?;
+        let (to_withdraw, to_slash) = amount
```
Comment (sander2) [inline crates/vault-registry/src/types.rs:454]: Shouldn't we use a saturating sub here? I.e. if the vault has _some_ collateral, but less than `amount`, should we not slash whatever it has?

Comment (gregdhill) [reply crates/vault-registry/src/types.rs:454]: I think this is already handled, if the vault has less collateral than the amount then the remainder is taken from the nominators. Implicitly the `amount` can not be greater than the collateral supplied by a vault and it's nominators.

Comment (sander2) [reply crates/vault-registry/src/types.rs:454]: Ah woops you are right

**Review by sander2 — APPROVED** [review]

**Pull request merged**